File: shannon_fano.c
#include <stdio.h>
#include "math.h"

//double numbers[6] = { 0.05, 0.05, 0.1, 0.2, 0.25, 0.35 };
double numbers[4] = { 0.1, 0.2, 0.3, 0.4 };

int code[255][255];
int split_index[100];

double count;
double err1 = 0;
double err2 = 0;

//sizeof(numbers)/sizeof(double)
int ind_start;
int ind_end;
int split_at;
int found_middle = 0;
int colum = 0;
int stack[255][3];
int stack_count = 0;
void find_middle() {
    count = 0;
    for (int i = ind_start; i < ind_end; i++) {
        if (found_middle) {
            code[i][colum] = 0;
        } else {
            code[i][colum] = 1;
        }
        //printf("%.2lf\n", numbers[i]);
        count += numbers[i];

        if (count >= 0.5 && !found_middle) {
            err1 = count - 0.5;
            err2 = fabs((count - numbers[i]) - 0.5);

            if (err1 > err2) {
                split_at = i;
            } else {
                split_at = i+1;
            }

            found_middle = 1;
            //printf("%.2lf\n", err1);
            //printf("%.2lf\n", err2);
            //printf("%d\n", split_index[0]);
        }
        
    }

    // Push next iteration
    stack_count++;
    colum++;

    stack[stack_count][0] = ind_start;
    stack[stack_count][1] = split_at - 1;
    stack[stack_count][2] = colum;

    stack[stack_count][0] = split_at;
    stack[stack_count][1] = ind_end;
    stack[stack_count][2] = colum;

    // Update variables
}


int main(int argc, char *argv[])  {

    ind_start = 0;
    ind_end = sizeof(numbers)/sizeof(double);

    find_middle();

    for (int i = 0; i < (sizeof(numbers)/sizeof(double)); i++) {
        for (int j = 0; j < 5; j++) {
            printf("%d", code[i][j]);    
        }  
        printf("\n");
    }  


    return 0;
}

File: shannon_fano.py
from Basic import *

global_shanon_fano = [        
        assign("A", 0),                 # count
        assign("B", 0),                 # err1
        assign("C", 0),                 # err2
        assign("D", 0),                 # sum

        alloc_matrix("B", 255, 3),      # stack
        assign("E", 0),                 # stack-count

        assign("F", 1),                 # ind_start
        assign("G", 1),                 # ind_end
        assign("H", 1),                 # column

        assign("I", 0),                 # split-at
        assign("J", 0),                 # found_middle
]
shannon_fano = main(
    pr_txt("Shannon Fano"),

    # number of codewords
    pr_txt("N:"),
    rd("N"),

    alloc_matrix("C", 1, "N"),      # codeword-list
    alloc_matrix("A", "N", 255),    # encoded matrix

    # read codeword probability
    cb_for("I", 1, "N", 1,
        pr_txt("enter p(i):"),
        rd("List 1[I]"),
    ),

    # sort ascending
    sortA("List 1"),

    # Init first iteration
    assign("E", "E+1"),
    assign("Mat B[E,1]", 1),
    assign("Mat B[E,2]", "N"),
    assign("Mat B[E,3]", 1),

    # Encode while stack not empty
    lbl("1"),
    cb_while("E>0",
        # Update variables
        assign("F", "Mat B[E,1]"),
        assign("G", "Mat B[E,2]"),
        assign("H", "Mat B[E,3]"),

        # Reset vars
        assign("A", 0),
        assign("J", 0),
        
        # Break condition
        cb_if(equal(sub("G", "F"), 1),
            _then=(
                assign("E", "E-1"),
                goto("1"),
            ),
        ),

        # Interval consist of two items
        cb_if(equal(sub("F", "G"), 1),
            _then=(
                assign("Mat A[F,H]", 1), # TODO: Fix dimenstion error (var F greater than matrix!)
                assign("Mat A[G,H]", 0),
                assign("E", "E-1"),
                goto("1"),
            ),
        ),

        # Get actual item sum
        assign("D", "Sum List 1"),

        cb_for("L", "F", "G", 1,
            cb_if("J=1",
                _then=(
                    assign("Mat A[L,H]", 0),
                ),
                _else=(
                    assign("Mat A[L,H]", 1),
                ),
            ),

            assign("A", "A+Mat C[1,L]"),

            cb_if(gt("A", mul("D", "0.5")+" And "+equal("J", 0)),
                _then=(
                    assign("B", sub("A", mul("D", "0.5"))),
                    assign("C", abslt(sub("A", "Mat C[1,L]", mul("D", "0.5")))),

                    cb_if(lore("B", "C"),
                        _then=( # TODO: fix if-else wrapper!
                            assign("I", "L"),
                        ),
                        _else=(
                            assign("I", "L-1"),
                            assign("Mat A[L,H]", 0),
                        ),
                    ),
                    assign("J", 1),
                ),
            ),
        ),

        # Push next iterations
        assign("H", "H+1"),

        # Dont increase stack to simulate 'pop' from stack
        assign("Mat B[E,1]", "F"),
        assign("Mat B[E,2]", "I"),
        assign("Mat B[E,3]", "H"),

        assign("E", "E+1"),
        assign("Mat B[E,1]", "I+1"),
        assign("Mat B[E,2]", "G"),
        assign("Mat B[E,3]", "H"),
    ),

    pr_var("Mat A"),

    globals=global_shanon_fano
)

if __name__ == '__main__':
    print(shannon_fano)